### ISSUE TITLE
Which dropout type is used for flash attention?

### ISSUE THREAD

**Issue by Avelina9X** [Dao-AILab/flash-attention#987]
I'm wondering which dropout type is utilised by the kernel version of flash attention. From the non flash implementation it looks like dropout is applied after softmax, so I assume this is the version used by the flash kernel?

However I just want to double check because other attention dropout implementations work by applying -inf to the dropped locations prior to softmax with some implementations also disabling dropout along the diagonal and/or reweighting the drop rate per row when operating in casual mode.

I'm asking because I want to ensure parity for the non flash version of my training code, and right now I'm throwing an error when dropout is enabled because I'm not sure which dropout implementation is used by the flash kernel.

**Comment by tridao** [comment]
The reference implementation is here: https://github.com/Dao-AILab/flash-attention/blob/320fb59487658f033f56711efd3d61b7c7a6f8f3/tests/test_flash_attn.py#L208
We apply dropout to the output of softmax.

**Comment by Avelina9X** [comment]
Excellent, thank you for the response!
